src/manifolds/SymplecticGrassmannStiefel.jl

```diff
@@ -0,0 +1,158 @@
+    \bigr),
+```
+where ``I_{2n}`` denotes the identity matrix and ``(⋅)^+`` the [`symplectic_inverse`](@ref).
+"""
+function inner(M::SymplecticGrassmann, p, X, Y)
+    n, k = get_parameter(M.size)
+    J = SymplecticElement(p, X, Y) # in BZ21 also J
+    # Procompute lu(p'p) since we solve a^{-1}* 3 times
+    a = lu(p' * p) # note that p'p is symmetric, thus so is its inverse c=a^{-1}
+    # we split the original trace into two one with I -> (X'Yc)
+    # 1) we permute X' and Y c to c^{\mathrm{T}}Y^{\mathrm{T}}X = a\(Y'X) (avoids a large interims matrix)
+    # 2) the second we compute as c (X'p)(p^+Y) since both brackets are the smaller matrices
+    return tr(a \ (Y' * X)) - tr(
+        a \ ((X' * p) * symplectic_inverse_times(SymplecticStiefel(2 * n, 2 * k), p, Y)),
+    )
+end
+
+@doc raw"""
+    inverse_retract(::SymplecticGrassmann, p, q, ::CayleyInverseRetraction)
+    inverse_retract!(::SymplecticGrassmann, q, p, X, ::CayleyInverseRetraction)
+
+Compute the Cayley Inverse Retraction on the Symplectic Grassmann manifold,
+when the points are represented as symplectic bases, i.e. on the [`SymplecticStiefel`](@ref).
+
+Here we can directly employ the `CaleyInverseRetraction` on the symplectic Stiefel manifold.
+"""
+inverse_retract(::SymplecticGrassmann, p, q, ::CayleyInverseRetraction)
+
+function inverse_retract_cayley!(M::SymplecticGrassmann, X, p, q)
+    n, k = get_parameter(M.size)
+    return inverse_retract_cayley!(SymplecticStiefel(2 * n, 2 * k), X, p, q)
+end
+
+@doc raw"""
+    retract(::SymplecticGrassmann, p, X, ::CayleyRetraction)
+    retract!(::SymplecticGrassmann, q, p, X, ::CayleyRetraction)
+
+Compute the Cayley retraction on the Symplectic Grassmann manifold,
+when the points are represented as symplectic bases, i.e. on the [`SymplecticStiefel`](@ref).
+
+Here we can directly employ the `CaleyRetraction` on the symplectic Stiefel manifold.
+"""
+retract(::SymplecticGrassmann, p, X, ::CayleyRetraction)
+
+function retract_cayley!(M::SymplecticGrassmann, q, p, X, t::Number)
+    n, k = get_parameter(M.size)
+    retract_cayley!(SymplecticStiefel(2 * n, 2 * k), q, p, X, t)
+    return q
+end
+
+@doc raw"""
+    riemannian_gradient(M::SymplecticGrassmann, p, Y)
+
+Given a gradient ``Y = \operatorname{grad} \tilde f(p)`` in the embedding ``ℝ^{2n×2k}`` or at
+least around the [`SymplecticGrassmann`](@ref) `M` where `p` (the embedding of) a point on `M`,
+and the restriction ``\tilde f`` to the [`SymplecticStiefel`] be invariant for the equivalence classes.
+In other words ``f(p) = f(qp)`` for ``q \in \mathrm{Sp}(2k, ℝ)``.
+Then the Riemannian gradient ``X = \operatorname{grad} f(p)`` is given by
+
+```math
+  X = J_{2n}^THJ_{2k}p^{\mathrm{T}}p - J_{2n}^TpJ_{2k}H^{\mathrm{T}}p,
+```
+
+where ``J_{2n}`` denotes the [`SymplecticElement`](@ref), and
+``H = (I_{2n} - pp^+)J_{2n}^{\mathrm{T}}YJ``.
+"""
+function riemannian_gradient(M::SymplecticGrassmann, p, Y; kwargs...)
+    n, k = get_parameter(M.size)
```

Comment:
Why do we need to implement both mutating and non-mutating variant?

Reply:
Then we could remove this, but I will only do that tomorrow (left my office already)